docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,421 @@
+        k8s.io/az: us-east-1
+```
+
+
+Later, based on these tags, the control-plane creates `ClusterLoadAssignment` and puts all tags, for each endpoint, into cluster metadata::
+
+```json
+"metadata":{
+  "filter_metadata": {
+    "envoy.lb": {
+      "az/test": "b",
+      "kuma.io/protocol": "tcp",
+      "k8s.io/az": "us-east-1"
+    }
+  }
+}
+```
+
+
+We can retrieve them from `ClusterLoadAssignment` and match them with the policy. First, we need to retrieve all tags from inbounds so we can determine our location and match it with endpoints.
+
+Pseudo algorithm:
+
+1. Get all inbound tags.
+2. Take the matched rules.
+3. Check if inbound has the tag defined in the rule. 
+4. If yes, iterate over all endpoints of the service and group them in to priority groups, by matching with the specific rule(first localZone, then  crossZone). If no, get all the endpoints in the zone.
+5. Create `LocalityLbEndpoints` with selected endpoints. 
+6. Override `ClusterLoadAssignment` for the dataplane in the `ResourceSet`.
+
+Dataplanes in the specific location are going to receive only endpoints in the specific locality. For cross zone traffic without egress the control-plane delivers only ingresses of specific zones.
+
+How to configure Clusters traffic?
+
+Envoy allows different forms of load balancing:
+* [Priority load balancing](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/priority.html#priority-levels)
+* [Locality weighted load balancing](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/locality_weight#arch-overview-load-balancing-locality-weighted-lb)
+* [Load Balancer Subsets](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/subsets.html)
+
+1. Priority: Ideal for managing cross-zone traffic and prioritizing the acceptance of specific zones. You can also use it to prioritize local zone endpoints based on these settings.
+
+2. Load Balancer Subsets: Initially considered a good choice for managing localZone traffic. It organizes endpoints into subsets based on provided keys, albeit with a slight increase in Envoy's memory usage. However, I encountered some issues during testing that led me to reconsider this option:
+
+Problems:
+* When an endpoint with a matching tag becomes unhealthy, Envoy still attempts to send requests to it and doesn't fall back. This behavior is unacceptable.
+* Priorities do not work seamlessly with subsets, even when using aggregated clusters.
+
+3. Locality Weighted Load Balancing: This strategy can help distribute traffic and may be suitable for local zone traffic, but it's not a perfect solution due to the following concerns:
+* We need to deliver weights to the sidecar, which can cause traffic to not stay within the node. Traffic is routed based on these weights, so even if we specify that a node has a higher weight than an availability zone, some traffic might still be routed elsewhere.
```

Comment:
I feel like it's good but not perfect. It might be difficult to choose the right weight and we are not giving the guarantee that 100% of the traffic stays in the `k8s.io/node` and some requests might get out of the node even if there are enough endpoints to handle it.

We have 2 options:
1. Use only priority
2. Use priority for cross-zone and Locality Weighted Load Balancing for the local zone with different tags. But that won't guarantee that we redirect all the traffic to the same node, but we are going to prefer it.

Reply:
I think we should implement second option for now

Reply:
But if you have 10 instances of the service in your zone and only 1 in the same node as you. Do you want to direct 100% of the traffic to this one instance?

I guess it's a tradeoff not clear what's the preferred option

Reply:
It's better to not DDos your service, the only thing that I didn't like with the Weighted Load balancing is that when there are not enough endpoints based on the overprovisioning factor, traffic might go to another zone. So even if you have 2 instances out of 3 running in zone an overprovisioning factor might cause some of the traffic can be routed cross zone.